autopilot/contractor.go

```diff
@@ -221,6 +233,129 @@ func (c *contractor) performContractMaintenance(cfg api.AutopilotConfig, cs api.
 	return c.ap.bus.SetContractSet("autopilot", contractset)
 }
 
+func (c *contractor) performWalletMaintenance(cfg api.AutopilotConfig, cs api.ConsensusState) error {
+	c.logger.Info("performing wallet maintenance")
+	b := c.ap.bus
+	l := c.logger
+
+	// no contracts - nothing to do
+	numHostWanted := cfg.Contracts.Hosts
+	if numHostWanted == 0 {
+		l.Debug("wallet maintenance skipped, no contracts wanted")
+		return nil
+	}
+
+	// no allowance - nothing to do
+	if cfg.Contracts.Allowance.IsZero() {
+		l.Debug("wallet maintenance skipped, no allowance set")
+		return nil
+	}
+
+	// no money - nothing to do
+	balance, err := b.WalletBalance()
+	if err != nil {
+		return err
+	}
+	if balance.IsZero() {
+		l.Debug("wallet maintenance skipped, zero balance in wallet")
+		return nil
+	}
+
+	// fetch pending transactions
+	pending, err := c.ap.bus.WalletPending()
+	if err != nil {
+		return nil
+	}
+
+	// pending maintenance transaction - nothing to do
+	c.mu.Lock()
+	currID := c.maintenanceTxnID
+	c.mu.Unlock()
+	for _, txn := range pending {
+		if currID == txn.ID() {
+			l.Debugf("wallet maintenance skipped, pending transaction found with id %v", currID)
+			return nil
+		}
+	}
+
+	// fetch wallet outputs
+	outputs, err := b.WalletOutputs()
+	if err != nil {
+		return err
+	}
+	numOutputs := uint64(len(outputs))
+
+	// enough outputs - nothing to do
+	if numOutputs > numHostWanted {
+		l.Debugf("no wallet maintenance needed, plenty of outputs available (%v>%v)", numOutputs, numHostWanted)
+		return nil
+	}
+
+	// fetch active contracts
+	active, err := b.ActiveContracts()
+	if err != nil {
+		return err
+	}
+	numContracts := uint64(len(active))
+
+	// if enough contracts - check if we have enough outputs for refreshes
+	var missingForRefresh int
+	var missingForFormation int
+	if numContracts >= numHostWanted {
+		numOutputsForRefresh := uint64(math.Ceil(float64(numContracts) * minOutputsForRefreshesPct))
+		if numOutputs >= numOutputsForRefresh {
+			return nil
+		}
+		missingForRefresh = int(numOutputsForRefresh - numOutputs)
+	} else {
+		missingForFormation = int(numOutputs) - int(numHostWanted-numContracts)
+		missingForFormation = int(math.Max(float64(missingForFormation), 0))
+	}
+
+	// enough outputs - nothing to do
+	missing := missingForFormation + missingForRefresh
+	if missing == 0 {
+		l.Debugf("no wallet maintenance needed, enough outputs available (%d)", numOutputs)
+		return nil
+	}
+
+	// calculate a good amount to redistribute
+	amount := balance.Div64(uint64(missing))
+	allowance := cfg.Contracts.Allowance.Div64(cfg.Contracts.Hosts)
+	avgInitialContractFunds := allowance.Div64((minInitialContractFundingDivisor + maxInitialContractFundingDivisor) / 2)
+	if amount.Cmp(avgInitialContractFunds) < 0 {
+		amount = avgInitialContractFunds
+	}
+
+	// not enough money left
+	if amount.Mul64(2).Cmp(balance) > 0 {
```

Comment:
Same in the code in this area. If the balance is smaller than the allowance, we could skip.
If it is more than the allowance, we try to create 50 (or however many) outputs with allowance/50 + a little bit value each.

That should be sufficient as long as the user keeps the wallet sufficiently funded which they should anyway. 

There is really no benefit in fine-tuning the outputs beyond that.